### PR TITLE
🚀 Instate documentation for MLJFlux

### PR DESCRIPTION
**Pull request by EssamWisam**
Includes API index and placeholders for the documentation structure.

### CODE REVIEW

**Comment by EssamWisam** [comment]
@ablaom this is now ready for a review. However, this includes *only* the pages/sections under the "Introduction" or "Interface" headlines in the sidebar (tutorials and workflow examples and contributing are still work in progress; even if I committed some work related to them).

**Comment by ablaom** [comment]
Perfect. I will likely review before the weekend.

**Comment by ablaom** [comment]
Maybe is doesn't matter, but I don't ordinarily track docs/Manifest.toml.

**Comment by ablaom** [comment]
Still to do:

- [ ] Add doc generation to CI

**Comment by ablaom** [comment]
While I will defer to your superior sense of aesthetics, I must admit a preference for serif fonts. I find them easier to read. I think screen resolutions are so fantastic these days that [old arguments](https://webflow.com/blog/serif-vs-sans-serif#:~:text=Easier%20readability%20in%20print%2C%20but,if%20they%20have%20low%20resolution.) against Serif fonts don't apply.  

But again, happy to defer to your judgement and taste.

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:4]:
```suggestion
A Julia package integrating deep learning Flux models with [MLJ](https://juliaai.github.io/MLJ.jl/dev/).

```

Although we might need to update it again later :wink:

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:14]:
```suggestion
    MLJFlux support is focused on fundamental and widely used deep learning models.  Sophisticated architectures or techniques such as online learning, reinforcement learning, and adversarial networks are currently beyond its scope. 
```

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:28]:
```suggestion
```

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:33]: Have you thought removing the comments like "`# 1. Load Data` and making them ordinary text or headings outside of the fenced code?

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:49]: We don't need to `fit!` if we're going to evaluate.
```suggestion
```

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:57]:
```suggestion
Notice that we were also able to define the neural network in a high-level fashion by only specifying the number of neurons in each hidden layer and the activation function. Meanwhile, `MLJFlux` was able to infer the input and output layer as well as use a suitable default for the loss function and output activation given the classification task. Notice as well that we did not need to implement a training or prediction loop as in `Flux`.
```

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:71]:
```suggestion
```

File: docs/src/index.md
Comment (ablaom) [inline docs/src/index.md:87]:
```suggestion
[`FastAI`](https://fluxml.ai/FastAI.jl/dev/FastAI@dev/doc/docs/introduction.md.html) also provides a high-level interface for interacting with `Flux` models. If you are working exclusively with neural network models, and are less interested with comparison/composition with other kinds of machine learning models, this may be more suitable choice. Otherwise, `MLJFlux` may be better, especially for the users already familiar with the MLJ interface.
```

File: docs/src/interface/Summary.md
Comment (ablaom) [inline docs/src/interface/Summary.md:111]:
```suggestion
That said, you can only use MLJ loss functions or metrics in evaluation meta-algorithms (such as cross validation) and they will work even if the underlying model comes from `MLJFlux`.
```

File: docs/src/interface/Custom Builders.md
Comment (ablaom) [inline docs/src/interface/Custom Builders.md:19]:
```suggestion
        return Chain(
            Dense(n_in, nn.n1, init=init),
            Dense(nn.n1, nn.n2, init=init),
            Dense(nn.n2, n_out, init=init),
            )
```

File: docs/src/interface/Custom Builders.md
Comment (ablaom) [inline docs/src/interface/Custom Builders.md:24]: Can we have a link to the table, which is not on this page. Maybe this works:

```suggestion
Note here that `n_in` and `n_out` depend on the size of the data (see
[Table 1](@ref Models).
```

File: docs/src/interface/Custom Builders.md
Comment (ablaom) [inline docs/src/interface/Custom Builders.md:27]: Maybe you want a different link here?

```suggestion
the [Image Classification Example](@ref).
```

File: docs/src/interface/Custom Builders.md
Comment (ablaom) [inline docs/src/interface/Custom Builders.md:49]: Fixing some whitespace which is causing incorrect rendering (please check):

```suggestion
- `chain(x)` must make sense:
  - for any `x <: Array{<:AbstractFloat, 2}` of size `(n_in,
    batch_size)` where `batch_size` is any integer (for use with one
    of the first three model types); or
  - for any `x <: Array{<:Float32, 4}` of size `(W, H, n_channels,
    batch_size)`, where `(W, H) = n_in`, `n_channels` is 1 or 3, and
    `batch_size` is any integer (for use with `ImageClassifier`)
```

File: docs/src/interface/Custom Builders.md
Comment (ablaom) [inline docs/src/interface/Custom Builders.md:54]:
```suggestion
Alternatively, use [`MLJFlux.@builder(neural_net)`](@ref) to automatically create a builder for
```

File: docs/src/workflow examples/Incremental Training.md
Comment (ablaom) [inline docs/src/workflow examples/Incremental Training.md:45]:
```suggestion
curve = learning_curve(
    clf, X, y;
    range=r,
    resampling=Holdout(fraction_train=0.7),
    measure=cross_entropy,
    )
```

File: docs/src/workflow examples/Incremental Training.md
Comment (ablaom) [inline docs/src/workflow examples/Incremental Training.md:51]:
```suggestion
plot(
    curve.parameter_values,
    curve.measurements,
    xlab=curve.parameter_name,
    xscale=curve.parameter_scale,
    ylab = "Cross Entropy",
    )
```

File: docs/src/workflow examples/Incremental Training.md
Comment (ablaom) [inline docs/src/workflow examples/Incremental Training.md:55]: In my preview, this image is not rendering, but I guess that's expected. Maybe we should move the image into doc/assets.

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:19]: Those dastardly tabs that render differently than in source file editor! I wonder if Documenter can fix this?

```suggestion
    return reshape(x, :, size(x)[end])
```

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:46]:
```suggestion
	front = Chain(
            Conv((k, k), n_channels => c1, pad=(p, p), relu),
            MaxPool((2, 2)),
            Conv((k, k), c1 => c2, pad=(p, p), relu),
            MaxPool((2, 2)),
            Conv((k, k), c2 => c3, pad=(p, p), relu),
            MaxPool((2 ,2)),
           flatten,
           )
```

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:84]:
```suggestion
clf = ImageClassifier(
    builder=MyConvBuilder(3, 16, 32, 32),
    epochs=10,
    loss=Flux.crossentropy,
    )
```

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:95]:
```suggestion
evaluate!(
    mach,
    resampling=Holdout(rng=123, fraction_train=0.7),
    measure=misclassification_rate,
    )
```

File: docs/make.jl
Comment (ablaom) [inline docs/make.jl:47]: I guess we don't need this?

```suggestion
```

**Review by ablaom — COMMENTED** [review]
Thanks @EssamWisam . I've finished my initial review.

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:21]: I'm working on an independent PR to update MNIST and I noticed the following changes are needed for latest versions of packages, to avoid some warnings. 

So:

- Get rid `flatten` definition. Instead `import MLUtils` and use `MLUtils.flatten` instead.
- Change way data is loaded as noted in suggestion below.

Will copy you into the PR when it's ready.

File: docs/src/full tutorials/MNIST.md
Comment (ablaom) [inline docs/src/full tutorials/MNIST.md:56]:
```suggestion
Xraw, yraw = MNIST(split=:train)[:];
```

File: docs/src/index.md
Comment (EssamWisam) [reply docs/src/index.md:48]: Oops. Such a rookie mistake. Thanks. 

P.S.: Need to modify the comment as well.

File: docs/src/index.md
Comment (EssamWisam) [reply docs/src/index.md]: Somehow I remember removing this before making PR ready for review 🤷.

File: docs/src/index.md
Comment (EssamWisam) [reply docs/src/index.md:83]: @ablaom while I have been only browsing and searching in the docs for the last few minutes, my conclusion so far is that `FastAI` does not go too much beyond `providing a high-level interface for interacting with Flux models. `

For instance, I could neither by searching their docs nor quickly browsing the pages find that they support features such as advanced out-of-sample evaluation (e.g., cross validation) or hyperparameter search (e.g., tuning the number of hidden layers/number of neurons per layers). [FluxTraining.jl](https://fluxml.ai/FluxTraining.jl/dev/documents/README.md) (related to FastAI in some way) seems to only support hyperparameter scheduling which is not exactly hyperparameter search (e.g., still can't tune the number of layers with this; can argue it's not really "tuning" whatsoever). That same package does support early stopping though.

Thus, I find the following to not the most accurate fact, in that `FastAI` is not a drop-in replacement for `MLJFlux`:

> If you are working exclusively with neural network models, and are less interested with comparison/composition with other kinds of machine learning models, this (i.e., FastAI) may be more suitable choice.

I'm also not sure why the mention `exclusively with neural network models` because in their README they tend to just speak more generically of deep learning models.


I would recommend one of the following line of actions:
- That we don't mention `FastAI` at all (i.e., what an average package would do; also avoids confusion to users that: (i), only know `MLJFlux` and  (ii), where `MLJFlux` is sufficient for them; if (ii) did not hold then they would have likely been able to find FastAI themselves)
- Go out of our way and maximize sincerity as you intended and mention what we know about `FastAI` (e.g., after the comparison with Flux we add a line like `A comparable project, FastAI/FluxTraining, also provides a high-level interface for interacting with Flux models and supports a subset of features that may overlap with (but not include all of) those supported by MLJFlux`

For the last point, one can also go into more detail in the comparison but I don't find this necessary as both packages are likely to change with time and users may be in general less interested in such comparison. Comparing `MLJFlux` and `Flux` makes lots of sense because it's guaranteed everyone likely knows `Flux` and this helps them understand `MLJFlux` better.

File: docs/src/workflow examples/Incremental Training.md
Comment (EssamWisam) [reply docs/src/workflow examples/Incremental Training.md:53]: I supposedly did not intend for pages in `Workflow Examples` and `Tutorials` to be ready for review as I mentioned as I will likely revisit them; however, these review bit do make it clear to me to ensure formatting when I go back and modify or add examples.

File: docs/make.jl
Comment (EssamWisam) [reply docs/make.jl:47]: Yea.

**Comment by EssamWisam** [comment]
> Thanks @EssamWisam . I've finished my initial review.

Thank you so much for your time and effort in this fantastic review, @ablaom. I think it would make it easier for me to automate the documentation deployment if we merge this (after finishing with the remaining comments and removing any pages yet to be added; e.g., like in workflow examples) and then after merging make another PR for tutorials/workflow examples.

The idea is that I will want the action to trigger the deploy "on PR to dev" and for it to take the effect docs must be deployable from dev. It's likely though that there is a way to do this without merging first but I'm not sure of it, just yet.

**Comment by EssamWisam** [comment]
> While I will defer to your superior sense of aesthetics, I must admit a preference for serif fonts. I find them easier to read. I think screen resolutions are so fantastic these days that [old arguments](https://webflow.com/blog/serif-vs-sans-serif#:~:text=Easier%20readability%20in%20print%2C%20but,if%20they%20have%20low%20resolution.) against Serif fonts don't apply.
> 
> But again, happy to defer to your judgement and taste.

If I remember correctly, I tried to copy the font used in `Flux` documentation just for uniformity. While I didn't even know about the screen resolution argument (which is now broken anyway); I do think that using serifs in a site makes go from feeling more modern and contemporary to more classical and formal. Maybe if we adopted a less fancy logo and put some equations in the homepage, I'd feel better with serif, but I'm not much drawn to it in this case. Likewise, I do think that people on average (esp. Gen Z kids like me) would think the same.

I know you didn't require this much justification but I also wanted to clarify why I think what I think.

**Comment by ablaom** [comment]
> I know you didn't require this much justification but I also wanted to clarify why I think what I think.

Appreciate it! Yes, this may well be a generational thing, and I support your decision to keep the "modern" fonts.

File: docs/src/index.md
Comment (ablaom) [reply docs/src/index.md:32]: Got it! Makes sense.

File: docs/src/index.md
Comment (ablaom) [reply docs/src/index.md:83]: I think it's helpful to make links to other related packages. I also admit, that (out of some laziness) I may have been too generous is stating what FastAI.jl can do. That said, I have had some interactions with the developers and indirectly with the project and have the impression it is a well thought-out design by experts, and I remember a lot of community engagement. If I were working exclusively with images today - and certainly video - I expect I might prefer FastAI to MLJFlux.  

So the second suggestion sounds good, which I agree is more accurate.

**Review by ablaom — APPROVED** [review]
Great to see this progress. Merge when ready.

File: docs/src/index.md
Comment (EssamWisam) [reply docs/src/index.md:83]: Alright. I agree with everything you said. So we will go with replacing the text above with:
> A comparable project, FastAI/FluxTraining, also provides a high-level interface for interacting with Flux models and supports a set of features that may overlap with (but not include all of) those supported by MLJFlux.

Right? Let me know if you would to fine-tune it further.

Comment (ablaom) [reply docs/src/index.md:83]: all good

File: docs/src/full tutorials/MNIST.md
Comment (EssamWisam) [reply docs/src/full tutorials/MNIST.md:21]: I have hidden this tutorial anyway. Next PR will be focused on workflow examples and full tutorials

**Comment by EssamWisam** [comment]
@ablaom I can't merge because I don't have write access. Could you do instead? I added the script for deployment but it may be the case that someone with access to the repo needed to go to `settings/pages` and ensure deploying is activated.

**Comment by ablaom** [comment]
@EssamWisam You should have an invite for access now.

**Pull request merged**
